staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go

```diff
@@ -174,20 +179,44 @@ func NewSharedInformerFactoryWithOptions(client {{.clientSetInterface|raw}}, def
 	return factory
 }
 
-// Start initializes all requested informers.
 func (f *sharedInformerFactory) Start(stopCh <-chan struct{}) {
-  f.lock.Lock()
-  defer f.lock.Unlock()
+	f.lock.Lock()
+	defer f.lock.Unlock()
 
-  for informerType, informer := range f.informers {
-    if !f.startedInformers[informerType] {
-      go informer.Run(stopCh)
-      f.startedInformers[informerType] = true
-    }
-  }
+	if f.shutdownComplete {
+		return
+	}
+
+	for informerType, informer := range f.informers {
+		if !f.startedInformers[informerType] {
+			f.wg.Add(1)
+			// We need a new variable in each loop iteration,
+			// otherwise the goroutine would use the loop variable
+			// and that keeps changing.
+			informer := informer
+			go func() {
+				defer f.wg.Done()
+				informer.Run(stopCh)
+			}()
+			f.startedInformers[informerType] = true
+		}
+	}
+}
+
+func (f *sharedInformerFactory) Shutdown() {
+	// Holding the lock while waiting ensures that:
+	// - only a single gorouting enters Wait, as required by WaitGroup
+	// - concurrent calls to Start and Shutdown wait until the
+	//   shut down is complete
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+
+	// Will return immediately if there is nothing to wait for.
+	f.wg.Wait()
+	f.shutdownComplete = true
```

Comment:
This can be "shuttingDown", and you can set it to true at the beginning. Then you don't have to hold the lock the whole time, and "Start" likewise can return immediately instead of blocking until Shutdown releases the lock.

Reply:
I had considered that, but then the behavior of Shutdown changes slightly: when one goroutine is blocked in Shutdown waiting for goroutines, then a second call to it in another goroutine has to return immediately because there is nothing it can wait on. It cannot block on the mutex (not locked while waiting), and it cannot call `f.wg.Wait` (may only be called by one goroutine at a time).

Is that okay?

Reply:
> cannot call f.wg.Wait (may only be called by one goroutine at a time).

Are you sure? I don't see that in the godoc for Wait()? I have let multiple goroutines call it at the same time in various places.

Reply:
I saw this in the source code:

https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/sync/waitgroup.go;l=140-142

> panic("sync: WaitGroup is reused before previous Wait has returned")

Can that be ignored?

Reply:
I tried to trigger that panic, but without success. Perhaps I was too cautious and read more into that panic than is warranted. I'm not sure what is meant with "reused". :shrug:

Reply:
My guess is that's internal housekeeping, but if it's not we should still be fine, since setting the flag & checking it before adding will prevent any more calls of Add, so there's no chance of racing between Add and Wait. Right?

Reply:
Agreed. I've updated the PR to use `shuttingDown` and not hold the mutex while waiting for goroutines.